plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -912,13 +916,24 @@ private AmazonS3 createAmazonS3Client(Configuration hadoopConfig, ClientConfigur
         }
 
         if (!regionOrEndpointSet) {
-            clientBuilder.withRegion(US_EAST_1);
+            Region region = Region.getRegion(US_EAST_1);
+            setRegion(region, clientBuilder, encryptionMaterialsProvider);
```

Comment:
I don't believe this specific scenario will actually work as intended. With S3, if no region or endpoint is set we set the region to `us-east-1` and then call `clientBuilder.setForceGlobalBucketAccessEnabled(true)` which in the case of S3 is special because S3's us-east-1 endpoint supports resolving the _actual_ region of a bucket by doing a "HEAD bucket" call. When `setForceGlobalBucketAccessEnabled(true)` is set, any S3 operation first checks a small cache of recently resolved bucket regions or resolves it via a head bucket call and then issues the request against that specific region.

If you set the region on the KMS client, however- you won't get that same behavior because us-east-1 is _not_ special there. However, we _do_ have the key ARN in the exception message which includes the region so it seems like it _should_ be possible to do something similar in principle although I might be missing something. We'll have to dig into the S3 encryption client logic to see if there's a similar mechanism to resolve the KMS region dynamically from the key ARN or whether we should just blow up when encryption materials are present but we're trying to use a global S3 access mode.

Reply:
I've tested multiple constellations (see https://github.com/trinodb/trino/pull/13715#issue-1341665075) and I find the functionality to work as expected.


The only time when I find the exception rather strange is while testing on `us-east-1` and pointing via `kmsKeyId` to a KMS key placed in `eu-central-1` availability time zone and in `hive.properties` `hive.s3.pin-client-to-current-region=true`. When doing an `INSERT` I receive :

```
io.trino.plugin.hive.s3.TrinoS3FileSystem$UnrecoverableS3OperationException: com.amazonaws.services.s3.model.AmazonS3Exception: Bad Request (Service: Amazon S3; Status Code: 400; Error Code: 400 Bad Request; Request ID: 0BDZW8G4C82S2S0F; S3 Extended Request ID: 0HLiuSFa5iS5UWb5ol76nT+WKRFXki/zlcVUVLeHmYyXlZy0zus+sBeh/hpg342HrKQpGkR9b58=; Proxy: null), S3 Extended Request ID: 0HLiuSFa5iS5UWb5ol76nT+WKRFXki/zlcVUVLeHmYyXlZy0zus+sBeh/hpg342HrKQpGkR9b58= (Path: s3://findinpath-sep-9566-eu-central-1/test22/20220818_080355_00000_hk6bv_2495dfb5-548b-432b-96c3-d9146ebc3f81)
	at io.trino.plugin.hive.s3.TrinoS3FileSystem.lambda$getS3ObjectMetadata$6(TrinoS3FileSystem.java:797)
```

Reply:
I've talked offline with @pettyjamesm and noticed that there is currently no way to explicitly specify in the s3 mappings file the region for a KMS key. 

When the  region or endpoint are not specified explicitly, but we want to use encryption, there is no way to know for KMS client where to retrieve the region for the KMS key ID specified. Probably in such a situation we should fail fast.
Note that the KMS key can be found in the same region as the s3 bucket or in a completely different region.
cc @losipiuk @electrum 


TODO Also the verify whether setting the signing region to `null` is correct/effective for retrieving the kms key.

```
        String endpoint = hadoopConfig.get(S3_ENDPOINT);
        if (endpoint != null) {
            clientBuilder.setEndpointConfiguration(new EndpointConfiguration(endpoint, null));
            regionOrEndpointSet = true;
        }
```